src/sonic-config-engine/minigraph.py

```diff
@@ -1192,6 +1252,25 @@ def parse_xml(filename, platform=None, port_config_file=None, asic_name=None, hw
         current_device['sub_role'] = sub_role
         results['DEVICE_METADATA']['localhost']['sub_role'] =  sub_role
         results['DEVICE_METADATA']['localhost']['asic_name'] =  asic_name
+    # on Voq system each asic has a switch_id
+    if switch_id is not None:
+        results['DEVICE_METADATA']['localhost']['switch_id'] = switch_id
+    # on Voq system each asic has a switch_type
+    if switch_type is not None:
+        results['DEVICE_METADATA']['localhost']['switch_type'] = switch_type
+    # on Voq system each asic has a max_cores
+    if max_cores is not None:
+        results['DEVICE_METADATA']['localhost']['max_cores'] = max_cores
+    # on Voq system each asic has a instance_name
+    if instance_name is not None:
+        results['DEVICE_METADATA']['localhost']['instance_name'] = instance_name
```

Comment:
We do not use the "instance_name" anywhere in the voq implementation. If this is not used anywhere we can remove this.
But we need "slot_id" and "device" in DEVICE_METADATA to derive the system lag name in the same way how system port name is derived (Refer: Voq LAG HLD https://github.com/Azure/SONiC/pull/697)

Reply:
I think "instance_name" here has the same meaning as "asic_name" in the VoQ LAG HLD, so I will rename "instance_name" -> "asic_name".

Reply:
I noticed that there is an api in minigraph.py that extracts asic_id from asic_name. This indicates that asic_name will be either of the two forms like asic0, asic1, etc. or asic<PCI device id> form.
If we have to have the asic_id in the format PCI device id, I do not think it will be good idea to use it for system lag name.

Therefore, I recommend to change the voq lag hld to use "slot_id" and "device" and add these attributes in device meta data so that both system port name and system lag name will be derived in similar way (using slot_id and device). We can get rid of "instance_name".

If this is fine with all, I'll take care of updating voq lag hld.

Reply:
I'm not sure that `asic_id` in minigraph.py has the same meaning as `asic_id` from CONFIG_DB. The former is used to extract the instance ID as described by systemd services, while the later is intended to be used internally by the platform to identify a particular asic (like PCI device ID, see also https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-orchagent/orchagent.sh#L28-L42).

I think `asic_id` in minigraph.py is misleading as it really is the "instance name" as defined by systemd.unit manual.

In any cases, using `slot_id` and `device` looks reasonable to avoid the confusion.

Reply:
Hi @vganesan-nokia 

It looks like `asic_name` in `minigraph.py` is actually the asic's hostname. Adding `device` to minigraph.xml will basically replicate asic's `Hostname`. I think it would be fair here to simply do:
```python3
        results['DEVICE_METADATA']['localhost']['asic_name'] = asic_name
```
or, if it makes more sense to name it `device_name` in CONFIG_DB:
```python3
        results['DEVICE_METADATA']['localhost']['device_name'] = asic_name
```

Let me know what you think.

Reply:
> I'm not sure that `asic_id` in minigraph.py has the same meaning as `asic_id` from CONFIG_DB. The former is used to extract the instance ID as described by systemd services, while the later is intended to be used internally by the platform to identify a particular asic (like PCI device ID, see also https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-orchagent/orchagent.sh#L28-L42).
> 
> I think `asic_id` in minigraph.py is misleading as it really is the "instance name" as defined by systemd.unit manual.

Instead of saying misleading, I would say, the minigraph.py does not support PCI device id format for asic_id. Currently minigraph.py takes the asic_id from asic_name, which is assumed to be in the format asic0, asic1, and so on. DEVICE_METATDATA['localhost']['asic_id'] is required attribute for multi-asic. This is actually passed to orchangent and also to SAI for device identification in SAI/sdk. As you mentioned this is also used for instance id (namespace) in systemd services. So as long the asic_id has the name in the format "asic0", "asic1" etc we are good.  But since minigraph.py does not have support for handling and passing asic_id in PCI device id format, If we have to pass the asic_id in PCI device id format in config_db, we'll have to over-write this already filled asic_id by value we get from default_config.json (as being discussed in https://github.com/Azure/sonic-utilities/pull/1228). If this is going to be the case, we do not have to worry about asic_id in minigraph.

Reply:
> Hi @vganesan-nokia
> 
> It looks like `asic_name` in `minigraph.py` is actually the asic's hostname. Adding `device` to minigraph.xml will basically replicate asic's `Hostname`. I think it would be fair here to simply do:
> 
> ```python
>         results['DEVICE_METADATA']['localhost']['asic_name'] = asic_name
> ```
> 
> or, if it makes more sense to name it `device_name` in CONFIG_DB:
> 
> ```python
>         results['DEVICE_METADATA']['localhost']['device_name'] = asic_name
> ```
> 
> Let me know what you think.

I think both in minigraph and config_db "asic_name" takes the value "asic0", "asic1" and so on. Also note that asic_name is nothing but the namespace passed to parse_xml() by sonic-cfggen. And we have a separate config_db attribute for "hostname" which is "Hostname" in minigraph. So I guess "asic_name" cannot be hostname but should be considered as asic instance name. Therefore, we do not need "InstanceName" and "Device" in minigraph instead use the "asic_name".
 
The "hostname" which is used for both single asic and multi-asic should be equivalent to the line card in chassis context. So we can consider the "Hostname" for "slot_id" and avoid introducing "SlotId" in minigraph.

So my recommendations are:
(1) Set config db ['DEVICE_METADATA']['localhost']['asic_name'] = asic_name, as you suggested.
(2) Set config_db ['DEVICE_METADATA']['localhost']['hostname'] = hostname
(3) Remove "InstanceName" and use "asic_name" for instance name if instance name is needed
(4) For system port table: Remove "SlotId" and "Device". Use "hostname" and "asic_name" instead and derive system port key using the format "hostname|asic_name|portname"

I think (1) and (2) are already there as part of multi-asic. I feel that it is better to extend the use of existing multi-asic attributes for voq chassis systems also instead of introducing new attributes in config db and minigraph.

Reply:
Correct. We need "hostname" and "asic_name" in the DEVICE_METADATA and should use these for deriving the system port name something like  "<hostname>|<asic_name>|Ethernet1". The recommendation (4) is to say that we do not need to define "SlotId", 'Device" and "InstanceName" in minigraph. We already have "hostname" and "asic_name" and "asic_name" is same as instance name.

Reply:
This is sounds good to me. It will be good to have the values of HostName and AsicName consistent with values in DEVICE_METADATA "hostname" and "asic_name". The reason is: the values of  ['DEVICE_METADATA']['localhost']['hostname']  and  ['DEVICE_METADATA']['localhost']['asic_name'] are used for deriving the system lag name. It will be good to have naming convention consistent for system port also. So for here, if we make sure HostName and AsicName, use the same corresponding values (the values of  ['DEVICE_METADATA']['localhost']['hostname']  and  ['DEVICE_METADATA']['localhost']['asic_name'], it will be great.